Munge pr since last loop

apelisse opened the pull request:
I've tentatively explain what this does in the commit messages, please read those instead.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1819)

<!-- Reviewable:end -->

apelisse commented:
This is an attempt to improve the situation described by #1406

eparis commented on mungegithub/github/timer.go:
```diff
@@ -0,0 +1,52 @@
+
+	for id, date := range timer.ids {
+		if timer.now().After(date) {
+			delete(timer.ids, id)
```
This is likely fine (I just started reviewing) but it makes me feel icky that `Get` has in implicit `Remove`.

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,77 @@
+
+// StateChange keeps track of issue/commit for state changes
+type StateChange struct {
+	issues  map[string]map[int]struct{} // Issues for each commit
```
Should we be using https://github.com/kubernetes/kubernetes/tree/master/pkg/util/sets

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,77 @@
+type StateChange struct {
+	issues  map[string]map[int]struct{} // Issues for each commit
+	commits map[int]string              // Commit for each issue
+	changed map[int]struct{}            // Changed issues since last call
```
When I see these I think of sets...

apelisse replied:
Yeah, agreed, I should be using set, will replace

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,77 @@
+// StateChange keeps track of issue/commit for state changes
+type StateChange struct {
+	issues  map[string]map[int]struct{} // Issues for each commit
+	commits map[int]string              // Commit for each issue
```
Is this just tracking the last commit? Not all of the commits?

eparis reviewed (commented):
I don't get the whole idea. My not understanding the github APIs in question. How does a 'commit' change? If I have a PR with HEAD=abcdef this will get registered in `commits` and `issues`. Great. Now lets say I edit my PR such that HEAD=123456. What is going to happen here?

commits will be updated to 123456.
issues will still have abcdef pointing to my PR, but it will not also have 123456 pointing to my PR.

When does `Change` get called? 

I'm going to go read the 3rd PR now, so this might all be non-sense.

Commit:
mungegithub: Only fetch issue changed since last loop

Use the "since" parameter from Github List Issue interface, so that we
don't munge issues that haven't changed since the last loop.

apelisse commented:
@eparis I'll try to clarify the idea

It looks like a Status change doesn't update the UpdateAt field of an issue/pr, which means that using the "Since" parameter, we won't get issues whose status as changed (but nothing else). And I assume this is a problem.

Listening to Status change webhooks event helps, but the problem is that it doesn't say what pull-requests are impacted by the status-change, just the commit sha. That's why I'm keeping track of "what issues are associated with what commit sha" so that I can say "Ah! This sha status has changed, it means these PRs have changed, we should probably also munge them."

apelisse commented:
Also, I only care about the HEAD because this is the status we are interested in.

Commit:
mungegithub: Track status change to munge more issues

We now use "since" to munge only issues that changed. Unfortunately,
Github doesn't track "state change" on commits as being a change to the
issue (because it tracks it per commit rather than per issue). Let's
track `issue <-> commit` ourselves so that we can munge issues whose
head commit state has changed.

apelisse commented:
@eparis I've updated the PR, and re-ordered the commits to maybe clarify the idea. I've also dropped the timer for now.

What this is doing:
- Use "Since" so we only munge issues that have changed since last loop
- Use webhooks and track StateChange so that we can munge issues whose HEAD commit status has changed.

Does it make more sense?

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,71 @@
+func (s *StateChange) Change(commit string) {
+	s.mutex.Lock()
+	s.changed.Insert(s.issues[commit].List()...)
+	s.mutex.Unlock()
```
Please use
```
lock()
defer unlock()
```
throughout.

apelisse replied:
Sure, why not.

eparis commented on mungegithub/mungegithub.go:
```diff
@@ -107,6 +107,8 @@ func main() {
 				glog.Fatalf("unable to initialize mungers: %v", err)
 			}
+			config.HookHandler.Initialize()
+			go config.HookHandler.Listen()
```
Didn't you have code to make it all work without turning the web hook on? Do we need such code? Is there a reason not to put the `Listen()` inside the `Initialize()`?

apelisse replied:
No, we need the webhooks to listen to status (unfortunately). I'll move the listen.

eparis commented on mungegithub/github/webhook.go:
```diff
@@ -0,0 +1,79 @@
+
+// Listen receives webhooks.
+func (webhook *WebHook) Listen() {
+	http.Handle("/", webhook)
```
Isn't re-using the `http.DefaultServeMux` here going to interfere with the usage for the submit queue UI on :80 ?

apelisse replied:
Very good catch

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// UpdateCommit must be called when we know how commits and issues are associated
+func (s *StateChange) UpdateCommit(issue int, commit string) {
```
Is this dead code now?

apelisse replied:
No, I just forgot to call it ..

eparis commented on mungegithub/github/state_change.go:
```diff
@@ -0,0 +1,71 @@
+// StateChange keeps track of issue/commit for state changes
+type StateChange struct {
+	issues  map[string]sets.Int // Issues for each commit
+	commits map[int]string      // Commit for each issue
```
If `UpdateCommit()` is not dead code, so is `commits` I think.

eparis commented on mungegithub/github/github.go:
```diff
@@ -1851,13 +1860,30 @@ func (config *Config) ForEachIssueDo(fn MungeFunction) error {
 	config.since = since
+
+	// Handle additional issues
+	for id := range extraIssues {
```
It'd be pretty awesome if we could do these async instead of waiting for the next loop, but this is certainly safest. I wonder how many of the mungers can handle getting called twice at the same time...

apelisse commented:
@eparis I tried to address your comment and give methods of StatusChange (that I renamed)  a better name.

Hopefully this goes in the right direction

apelisse commented:
@eparis I tried it with ngrok against kubernetes/kubernetes and fixed a couple of small bugs. This seems to be working now and definitely improves the token situation by orders of magnitude. Can we get that in?

Thanks

rmmh commented on mungegithub/github/github.go:
```diff
@@ -1800,46 +1834,30 @@ func (config *Config) ForEachIssueDo(fn MungeFunction) error {
 			Labels:      config.Labels,
 			Direction:   "asc",
 			ListOptions: github.ListOptions{PerPage: 100, Page: page},
+			Since:       config.since,
```
The JSON for this is `omitempty` -- why wouldn't zero time work?

rmmh commented on mungegithub/github/github.go:
```diff
@@ -1787,11 +1790,42 @@ func (obj *MungeObject) MergedAt() *time.Time {
+	if config.since.IsZero() {
+		config.since = time.Date(1900, 1, 1, 0, 0, 0, 0, time.UTC)
+	}
+	since := time.Now()
```
We may get some weirdness with clock skew here. Can we use the updated time of the last issue in the loop instead?

rmmh commented on mungegithub/github/github.go:
```diff
@@ -1800,46 +1852,47 @@ func (config *Config) ForEachIssueDo(fn MungeFunction) error {
 			}
-			if err := fn(&obj); err != nil {
-				continue
+			err := config.runMungeFunction(obj, fn)
```
there are mungers that need to run not just when things change -- like the stale issue nagging. How will this be handled?

Commit:
mungegithub: Stop assuming we'll receive each issues

The mungebot now only munges issues that have been modified since the
last loop. Some of the mungers made the assumptions that they would
receive all the issues, by refreshing their list for example. Try to
refresh the list in a better way so that we don't have to retrieve all
the list.

apelisse commented:
@eparis @rmmh Updated the PR to consider the following problem:

Some PR assume that they will receive the full list of open PRs (submit-queue, cherrypick-queue). I think I fixed this behavior.

apelisse commented:
Also, I realize this is currently being migrated to test-infra, I don't want to delay the review though. I'll diff/patch as soon as possible. But please feel free to start reviewing :-)
